compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -476,28 +476,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     .expr_ty_adjusted_opt(rcvr_expr)
                     .unwrap_or(Ty::new_misc_error(self.tcx)),
             );
-            /* err.span_note(
-                span,
-                format!("TEST: the method `{item_name}` is implemented on `{}`",
-                prev_ty.prefix_string(self.tcx)),
-            ); */
+            err.span_note(span, format!("{prev_ty:?}"));
+            println!("prev_ty: {}", prev_ty );
+            println!("item_name: {}", item_name);
 
-            if let Ok(_pick) = self.probe_for_name(
+            match self.probe_for_name(
             Mode::MethodCall,
                 item_name,
                 Some(prev_ty),
                 IsSuggestion(true),
                 prev_ty,
                 rcvr_expr.hir_id,
                 ProbeScope::AllTraits,) {
+                    Ok(_pick) => {
                     err.span_label(
                         span,
                         format!(
                         "{item_kind} `{item_name}` found for this `{}`",
                         prev_ty
                         ),
                  );
-                break;
+                },
+                Err(MethodError::NoMatch(..)) => 
+                    println!("No Match"),
+                Err(MethodError::Ambiguity(..)) =>  println!("Ambiguity"),
+                Err(MethodError::PrivateMatch(..)) => println!("Private"),
+                Err(MethodError::IllegalSizedBound { .. }) => println!("IllegalSizedBound"),
+                Err(MethodError::BadReturnType) => println!("BadReturnType"),
```

Comment:
This outputs:
```
prev_ty: &C
item_name: foo
No Match
prev_ty: &B
item_name: foo
BadReturnType
prev_ty: &A
item_name: foo
No Match
```
So, what makes the return type of `B` bad?

Reply:
@rustbot review

Reply:
I'm pretty sure it's because of this check:

https://github.com/rust-lang/rust/blob/dac91a82e146d89fefe94dcbb53cd6c2662e354f/compiler/rustc_hir_typeck/src/method/probe.rs#L1697-L1704

Can you take a look at what those two types in the `debug!` are, as well as the actual value of the error? It might be that we might want to consider `BadReturnType` as "correct" for our purposes (if we don't change that earlier code that produces it, potentially adding a new more specific  `ProbeResult` variant).

Reply:
Can we look at what the output of the following is?

https://github.com/rust-lang/rust/blob/dac91a82e146d89fefe94dcbb53cd6c2662e354f/compiler/rustc_hir_typeck/src/method/probe.rs#L1309

That might give us a clue at why we end up with a BadReturnType or NoMatch at the end of that function.

Reply:
BTW, you can prepend your rustc invocation with `RUSTC_LOG` like `RUSTC_LOG=rustc_hir_typeck::method::probe=debug rustc foo.rs` to make these debug statements be printed out when executed.

Reply:
I was going to ask.. 
A few other questions:
How do I run `stage1` compiler on gdb? I tried but just pointing to the stage1 binary crashed in gdb. Do I need to build `stage2` as well and point to that? 

What do I do to run an `ignored` test? - I am currently editing to ensure a recompile

Is there a specific place in the documentation and code that could help me understand the guts of this piece of analysis?

Reply:
@estebank Also wondering about the timeline expectations for PRs to be resolved - Do not want to take up all the time in learning on this PR if it's taking way too much of your time. But absolutely appreciate your time and look to learning more!

Reply:
> Is there a specific place in the documentation and code that could help me understand the guts of this piece of analysis?

In general, the best place to look at is the [dev guide](https://rustc-dev-guide.rust-lang.org/) for high level information, or the doc comments scattered throughout the different modules.

> What do I do to run an ignored test? - I am currently editing to ensure a recompile

What I do is [set up a rustup toolchain for my locally built](https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html?highlight=rustup#creating-a-rustup-toolchain) with the name `dev`, and after that calling `rustc +dev foo.rs` will compile `foo.rs`. Once you have that set up you can compile any of the tests (but this is only really useful for the tests that are self-contained, the ones that have dependencies are more annoying to deal with and it's easier to just run the tests for those).

What was the crash you saw with gdb? In general I rely heavily in debug statements instead of the debugger, but I've had good results with the lldb + vscode integration, so I don't know how to help you there.

> Also wondering about the timeline expectations for PRs to be resolved

There's no rush.

> Do not want to take up all the time in learning on this PR if it's taking way too much of your time. But absolutely appreciate your time and look to learning more!

Don't worry, and it's my pleasure. We all started somewhere.